File: pycore/callmodule/services/queue_center_contract.py
# -*- coding: utf-8 -*-
"""Canonical Queue Center and distributed-task contract adapter.

The language-neutral source is ``config/queue_center_contract.json``. Keep this adapter,
the Laravel ``App\Support\QueueCenterContract`` adapter, and the TypeScript
``core/api-libs/pycore/QueueCenterContract.ts`` adapter pointed at that file. The
mcp-chrome adapter is ``apps/mcp-chrome/app/chrome-extension/utils/queue-center-contract.ts``.

Changing a task status, execution lane, capability, wire field, or task-type route
must start in the JSON document. All four runtime adapters derive their values
from it; consumers must not introduce another literal vocabulary.
"""

import json
from copy import deepcopy
from pathlib import Path
from typing import Any, Dict, List, Literal, Mapping, Optional, Tuple, TypedDict

_PROJECT_ROOT = Path(__file__).resolve().parents[3]
_CONTRACT_PATH = _PROJECT_ROOT / "config" / "queue_center_contract.json"
_CONTRACT_DOCUMENT: Dict[str, Any] = json.loads(_CONTRACT_PATH.read_text(encoding="utf-8"))
_TASK_CONTRACT: Dict[str, Any] = _CONTRACT_DOCUMENT["task_contract"]

QueueCenterScope = Literal[
    "heartbeat",
    "assist_translation",
    "word_audio",
    "sentence_audio",
    "media_image",
]
QueueCenterSectionLifecycle = Literal["off", "starting", "on", "error"]


class QueueCenterToggleEnvelope(TypedDict):
    requested_by: Optional[str]
    enabled: bool
    reason: Optional[str]
    graceful_stop: bool
    paused_by_user: Optional[bool]


class QueueCenterControlMetrics(TypedDict):
    pending: int
    processing: int
    leased: int
    total: int


class QueueCenterWorkerMetrics(TypedDict):
    online: bool
    claimed: int
    ok: Optional[int]
    fail: Optional[int]
    last_heartbeat: Optional[str]


class QueueCenterSectionContract(TypedDict, total=False):
    type: QueueCenterScope
    category: str
    queue: QueueCenterControlMetrics
    worker: QueueCenterWorkerMetrics
    toggle: QueueCenterToggleEnvelope
    lifecycle: QueueCenterSectionLifecycle
    error_code: Optional[str]
    last_error: Optional[str]
    observed_at: Optional[str]
    age_s: Optional[float]
    stale: bool


class GlobalTaskRecord(TypedDict, total=False):
    """Language-neutral Laravel global-task wire model used by every consumer."""

    task_id: str
    app_name: str
    task_type: str
    execution_type: str
    capability: Optional[str]
    is_fast_tier: bool
    status: str
    priority: int
    progress: float
    payload: Dict[str, Any]
    result: Optional[Dict[str, Any]]
    error: Optional[str]
    assigned_to: Optional[str]
    timeout_seconds: int
    assigned_at: Optional[str]
    timeout_at: Optional[str]
    completed_at: Optional[str]
    created_at: Optional[str]
    updated_at: Optional[str]


QUEUE_CENTER_SCHEMA_VERSION = int(_CONTRACT_DOCUMENT["schema_version"])
QUEUE_CENTER_CONTROL_NAMES: Tuple[str, ...] = tuple(_CONTRACT_DOCUMENT["control_names"])
QUEUE_CENTER_CAPABILITY_CLAIMANTS: Dict[str, Tuple[str, ...]] = {
    key: tuple(str(item) for item in value)
    for key, value in _CONTRACT_DOCUMENT["capability_claimants"].items()
}
QUEUE_CENTER_SECTION_DEFINITIONS: Dict[str, Dict[str, Any]] = {
    key: dict(value)
    for key, value in _CONTRACT_DOCUMENT["section_scopes"].items()
}
QUEUE_CATEGORY_CATALOG: List[Dict[str, Any]] = []
for _category_definition in _CONTRACT_DOCUMENT["categories"]:
    _category = dict(_category_definition)
    _capability = _category.get("capability")
    _category["claimants"] = list(
        QUEUE_CENTER_CAPABILITY_CLAIMANTS.get(
            str(_capability),
            (str(_category["primary_handler"]),),
        )
    )
    QUEUE_CATEGORY_CATALOG.append(_category)
QUEUE_COUNT_KEYS: Tuple[str, ...] = tuple(_CONTRACT_DOCUMENT["metric_semantics"].keys())

GLOBAL_TASK_STATUSES: Tuple[str, ...] = tuple(_TASK_CONTRACT["statuses"]["all"])
GLOBAL_TASK_LIVE_STATUSES: Tuple[str, ...] = tuple(_TASK_CONTRACT["statuses"]["live"])
GLOBAL_TASK_TERMINAL_STATUSES: Tuple[str, ...] = tuple(_TASK_CONTRACT["statuses"]["terminal"])
GLOBAL_TASK_WORKER_RESULT_STATUSES: Tuple[str, ...] = tuple(
    _TASK_CONTRACT["statuses"]["worker_reportable"]
)
GLOBAL_TASK_EXECUTION_TYPES: Tuple[str, ...] = tuple(_TASK_CONTRACT["execution_types"])
GLOBAL_TASK_CAPABILITIES: Tuple[str, ...] = tuple(
    _CONTRACT_DOCUMENT["capability_claimants"].keys()
)
GLOBAL_TASK_PRIORITIES: Dict[str, int] = {
    str(key): int(value) for key, value in _TASK_CONTRACT["priorities"].items()
}
GLOBAL_TASK_CAPABILITY_SINGLE_LANES: Dict[str, str] = {
    str(key): str(value)
    for key, value in _TASK_CONTRACT["capability_single_lanes"].items()
}
GLOBAL_TASK_WIRE_SHAPES: Dict[str, Tuple[str, ...]] = {
    str(key): tuple(str(field) for field in value)
    for key, value in _TASK_CONTRACT["wire_shapes"].items()
}
GLOBAL_TASK_TYPE_CATALOG: Tuple[Dict[str, Any], ...] = tuple(
    dict(definition) for definition in _TASK_CONTRACT["task_types"]
)
GLOBAL_TASK_TYPES_BY_KEY: Dict[str, Dict[str, Any]] = {
    str(definition["key"]): definition for definition in GLOBAL_TASK_TYPE_CATALOG
}
_HISTORY_CONTRACT: Dict[str, Any] = _TASK_CONTRACT["history_buckets"]
GLOBAL_TASK_HISTORY_BUCKETS: Tuple[str, ...] = tuple(_HISTORY_CONTRACT["all"])

CALLBACK_QUEUE_ROLES: Dict[str, str] = {
    "translation_worker": "consumer",
    "translation_queue_monitor": "monitor",
    "translation_ws_client": "signal",
    "tts_queue_poller": "consumer",
    "tts_sentence_worker": "consumer",
    "ai_rate_reset": "maintainer",
    "agent_history_extraction": "maintainer",
    "agent_history_pipeline": "maintainer",
}


def get_queue_center_contract() -> Dict[str, Any]:
    """Return an isolated copy of the canonical JSON document."""
    return json.loads(json.dumps(_CONTRACT_DOCUMENT))


def task_type_definition(task_type: object) -> Optional[Dict[str, Any]]:
    """Return an isolated central task-type definition, if one exists."""
    key = str(task_type or "").strip().lower()
    definition = GLOBAL_TASK_TYPES_BY_KEY.get(key)
    return deepcopy(definition) if definition is not None else None


def task_execution_type(task_type: object, fallback: str = "remote_translation") -> str:
    """Resolve a Laravel task_type to its centrally assigned execution lane."""
    definition = task_type_definition(task_type)
    return str(definition.get("execution_type") or fallback) if definition else fallback


def task_capability(task_type: object) -> Optional[str]:
    """Resolve a task type's fixed capability; ``None`` means lane-only routing."""
    definition = task_type_definition(task_type)
    capability = definition.get("capability") if definition else None
    return str(capability) if capability else None


def task_types_for_execution(execution_type: str) -> Tuple[str, ...]:
    """Return task types assigned to one central execution lane."""
    return tuple(
        str(definition["key"])
        for definition in GLOBAL_TASK_TYPE_CATALOG
        if definition.get("execution_type") == execution_type
    )


def task_local_label(task_type: object, capability: object = None) -> str:
    """Resolve the pycore local TaskManager label for one Laravel task."""
    if str(capability or "") == "ai_translate":
        return "remote_ai_translate"
    definition = task_type_definition(task_type)
    if definition:
        return str(definition.get("pycore_local_label") or "remote_translation")
    return "remote_translation"


def project_task_record(raw: Mapping[str, Any], shape: str) -> GlobalTaskRecord:
    """Project an arbitrary mapping onto a central global-task wire shape."""
    fields = GLOBAL_TASK_WIRE_SHAPES.get(shape)
    if fields is None:
        raise ValueError(f"Unknown global-task wire shape: {shape}")
    return {field: raw.get(field) for field in fields}  # type: ignore[return-value]


def normalize_task_history_type(raw_task_type: object) -> str:
    """Map raw and legacy task names to the shared completed-history bucket."""
    fallback = str(_HISTORY_CONTRACT["fallback"])
    task_type = str(raw_task_type or "").strip().lower()
    if not task_type:
        return fallback
    if task_type in GLOBAL_TASK_HISTORY_BUCKETS:
        return task_type

    exact_aliases = _HISTORY_CONTRACT.get("exact_aliases", {})
    exact = exact_aliases.get(task_type)
    if exact:
        return str(exact)

    for rule in _HISTORY_CONTRACT.get("token_rules", []):
        all_tokens = tuple(str(token) for token in rule.get("all", []))
        any_tokens = tuple(str(token) for token in rule.get("any", []))
        if all(token in task_type for token in all_tokens) and (
            not any_tokens or any(token in task_type for token in any_tokens)
        ):
            return str(rule["bucket"])
    return fallback


def category_keys_for_scope(scope: QueueCenterScope) -> Tuple[str, ...]:
    raw = QUEUE_CENTER_SECTION_DEFINITIONS.get(scope, {}).get("category_keys", [])
    return tuple(str(value) for value in raw)


def build_empty_queue_contract(
    scope: QueueCenterScope,
    observed_at: Optional[str] = None,
) -> QueueCenterSectionContract:
    definition = QUEUE_CENTER_SECTION_DEFINITIONS[scope]
    contract: QueueCenterSectionContract = {
        "type": scope,
        "category": str(definition["category"]),
        "observed_at": observed_at,
    }
    contract.update(deepcopy(_CONTRACT_DOCUMENT["section_contract_defaults"]))
    return contract


__all__ = [
    "CALLBACK_QUEUE_ROLES",
    "QUEUE_CATEGORY_CATALOG",
    "QUEUE_CENTER_CONTROL_NAMES",
    "QUEUE_CENTER_CAPABILITY_CLAIMANTS",
    "QUEUE_CENTER_SCHEMA_VERSION",
    "QUEUE_CENTER_SECTION_DEFINITIONS",
    "QUEUE_COUNT_KEYS",
    "GLOBAL_TASK_CAPABILITIES",
    "GLOBAL_TASK_CAPABILITY_SINGLE_LANES",
    "GLOBAL_TASK_EXECUTION_TYPES",
    "GLOBAL_TASK_HISTORY_BUCKETS",
    "GLOBAL_TASK_LIVE_STATUSES",
    "GLOBAL_TASK_PRIORITIES",
    "GLOBAL_TASK_STATUSES",
    "GLOBAL_TASK_TERMINAL_STATUSES",
    "GLOBAL_TASK_TYPE_CATALOG",
    "GLOBAL_TASK_TYPES_BY_KEY",
    "GLOBAL_TASK_WIRE_SHAPES",
    "GLOBAL_TASK_WORKER_RESULT_STATUSES",
    "GlobalTaskRecord",
    "QueueCenterControlMetrics",
    "QueueCenterScope",
    "QueueCenterSectionContract",
    "QueueCenterSectionLifecycle",
    "QueueCenterToggleEnvelope",
    "QueueCenterWorkerMetrics",
    "build_empty_queue_contract",
    "category_keys_for_scope",
    "get_queue_center_contract",
    "normalize_task_history_type",
    "project_task_record",
    "task_capability",
    "task_execution_type",
    "task_local_label",
    "task_type_definition",
    "task_types_for_execution",
]

File: poly_apps/pycore_laravel_wordnew_ui/core/contracts/QueueCenterContract.ts
/**
 * TypeScript adapter for the canonical Queue Center and distributed-task contract.
 *
 * Source: config/queue_center_contract.json
 * Aligned adapters:
 * - pycore/pyutils/common/queue_center_contract.py
 * - poly_apps/laravel_main/app/Support/QueueCenterContract.php
 * - apps/mcp-chrome/app/chrome-extension/utils/queue-center-contract.ts
 *
 * This adapter is shared by Pycore UI and Laravel-manager UI. Change task
 * statuses, lanes, capabilities, task routes, priorities, and wire fields only
 * in the JSON source so Laravel, Pycore, mcp-chrome, and both UIs update together.
 */
import contractDocument from '../../../../config/queue_center_contract.json';

import type {
  QueueCenterControlName,
  QueueCenterScope,
  PcQueueHandler,
  QueueDeliveryStage,
  GlobalTaskStatus,
  GlobalTaskOrdering,
  GlobalTaskExecutionType,
  GlobalTaskCapability,
  GlobalTaskPayload,
  QueueCenterHttpTransfer,
  GlobalTaskCreateResult,
  GlobalTaskStatsRecord,
  GlobalTaskSummary,
  GlobalTaskWorkerRecord,
  GlobalTaskDetailRecord,
  GlobalTaskStatusRecord,
  GlobalTaskEventRecord,
  GlobalTaskCurrentPhase,
  GlobalTaskDetailMetadata,
  GlobalTaskDetailBundle,
  GlobalTaskWorkerRegistration,
  GlobalTaskWorkerResult,
  GlobalTaskTypeDefinition,
  GlobalTaskOrderingRecord,
  QueueCenterSectionContract,
} from './QueueCenterTypes';

export type * from './QueueCenterTypes';

interface ContractDocument {
  schema_version: number;
  http_transfer: QueueCenterHttpTransfer;
  realtime: {
    transport: string;
    topic: string;
    event: string;
    events: Record<string, string>;
  };
  diff_delivery: {
    version: number;
    cursor_store: string;
    id_page_store: string;
    data_segment_store: string;
    id_page_limit: number;
    id_limit: number;
    data_segment_limit: number;
    poll_interval_ms: number;
    head_reserve: number;
    head_notification_interval_seconds: number;
    producer_batch_limits: Record<string, number>;
    consumer_batch_limits: Record<string, number>;
    consumer_task_timeout_seconds: Record<string, number>;
    consumer_upload_retry: { initial_seconds: number; maximum_seconds: number };
    consumer_log_tags: Record<string, string[]>;
    ready_ttl_seconds: number;
    consumed_ttl_seconds: number;
  };
  endpoints: Record<string, string>;
  relay: {
    hub: {
      protocol: string;
      path: string;
    };
    device_identity: {
      clock_skew_seconds: number;
      nonce_ttl_seconds: number;
    };
    topics: Record<'machines' | 'pair', string>;
    events: Record<'request' | 'response' | 'roster', string>;
    machine_heartbeat_seconds: number;
    machine_offline_after_seconds: number;
    pair_session_ttl_seconds: number;
    request_ttl_seconds: number;
    response_poll_interval_ms: number;
    caps: {
      control_frame_bytes: number;
      inline_body_bytes: number;
      blob_chunk_bytes: number;
      request_total_bytes: number;
    };
    capability_providers: Record<string, {
      class: string;
      implemented: boolean;
      provides: string[];
    }>;
  };
  delivery_receipt: {
    stages: Record<string, QueueDeliveryStage>;
    worker_kinds: string[];
    task_id_limit_source: string;
  };
  control_names: QueueCenterControlName[];
  callback_queue_roles: Record<string, 'transport' | 'consumer' | 'monitor' | 'signal' | 'maintainer'>;
  section_contract_defaults: Omit<QueueCenterSectionContract, 'type' | 'category' | 'observed_at'>;
  capability_claimants: Record<string, PcQueueHandler[]>;
  section_scopes: Record<QueueCenterScope, {
    category: string;
    category_keys: string[];
    queue_metrics: boolean;
  }>;
  task_contract: {
    statuses: {
      values: Record<string, GlobalTaskStatus>;
      all: GlobalTaskStatus[];
      live: GlobalTaskStatus[];
      terminal: GlobalTaskStatus[];
      worker_reportable: GlobalTaskStatus[];
    };
    events: {
      values: Record<string, string>;
      terminal: string[];
    };
    stream_events: Record<'initial' | 'transition' | 'ping' | 'close', string>;
    execution_types: Record<string, GlobalTaskExecutionType>;
    priorities: Record<'default' | 'manual' | 'fast' | 'maximum', number>;
    progress_stages: Record<'accepted' | 'synthesizing' | 'uploading' | 'finalizing' | 'completed', number>;
    limits: Record<
      'list_default' | 'list' | 'monitor' | 'worker_pull_default' | 'worker_pull' | 'completed' | 'long_poll_seconds' | 'history_records' | 'history_timeline' | 'event_batch',
      number
    >;
    capability_labels: Record<GlobalTaskCapability, string>;
    capability_single_lanes: Record<GlobalTaskCapability, GlobalTaskExecutionType>;
    fast_lane_capabilities: GlobalTaskCapability[];
    wire_shapes: Record<string, string[]>;
    task_types: GlobalTaskTypeDefinition[];
    history_buckets: {
      all: string[];
      exact_aliases: Record<string, string>;
      token_rules: Array<{ bucket: string; all: string[]; any: string[] }>;
      fallback: string;
    };
  };
  categories: Array<{
    key: string;
    label: string;
    laravel_task_type: string | null;
    capability: string | null;
    primary_handler: PcQueueHandler;
  }>;
}

const GLOBAL_TASK_WIRE_DTO_FIELDS = {
  create_result: [
    'task_id', 'execution_type', 'queue_position', 'priority', 'is_fast_tier',
  ] as const satisfies readonly (keyof GlobalTaskCreateResult)[],
  summary: [
    'task_id', 'app_name', 'task_type', 'execution_type', 'status', 'progress',
    'assigned_to', 'created_at', 'capability', 'queue_position', 'priority', 'is_fast_tier',
  ] as const satisfies readonly (keyof GlobalTaskSummary)[],
  worker_pull: [
    'task_id', 'app_name', 'task_type', 'execution_type', 'status', 'payload',
    'timeout_seconds', 'retry_count', 'queue_position', 'priority', 'capability', 'is_fast_tier', 'created_at',
  ] as const satisfies readonly (keyof GlobalTaskWorkerRecord)[],
  status: [
    'task_id', 'app_name', 'task_type', 'execution_type', 'capability', 'is_fast_tier',
    'status', 'queue_position', 'priority', 'progress', 'retry_count', 'max_retries', 'timeout_seconds',
    'payload', 'result', 'error', 'assigned_to', 'assigned_at', 'timeout_at',
    'completed_at', 'created_at', 'updated_at',
  ] as const satisfies readonly (keyof GlobalTaskStatusRecord)[],
  detail: [
    'task_id', 'app_name', 'task_type', 'execution_type', 'capability', 'is_fast_tier',
    'status', 'queue_position', 'priority', 'progress', 'payload', 'result', 'error', 'assigned_to',
    'assigned_at', 'timeout_at', 'completed_at', 'created_at', 'updated_at',
  ] as const satisfies readonly (keyof GlobalTaskDetailRecord)[],
  event: [
    'id', 'task_id', 'event', 'worker_id', 'attempt', 'detail', 'created_at',
  ] as const satisfies readonly (keyof GlobalTaskEventRecord)[],
  detail_bundle: [
    'task', 'events', 'current_phase', 'metadata',
  ] as const satisfies readonly (keyof GlobalTaskDetailBundle)[],
  current_phase: [
    'phase', 'worker_id', 'elapsed_seconds',
  ] as const satisfies readonly (keyof GlobalTaskCurrentPhase)[],
  detail_metadata: [
    'total_attempts', 'max_retries', 'will_retry', 'estimated_timeout_in_seconds',
  ] as const satisfies readonly (keyof GlobalTaskDetailMetadata)[],
  stats: [
    'total', 'pending', 'assigned', 'processing', 'completed', 'completed_demo',
    'failed', 'cancelled',
  ] as const satisfies readonly (keyof GlobalTaskStatsRecord)[],
  worker_registration: [
    'worker_id', 'worker_name', 'processor_types', 'capabilities', 'hostname',
    'platform', 'metadata', 'lease_capacity',
  ] as const satisfies readonly (keyof GlobalTaskWorkerRegistration)[],
  worker_result: [
    'task_id', 'worker_id', 'attempt', 'status', 'progress', 'result', 'error',
  ] as const satisfies readonly (keyof GlobalTaskWorkerResult)[],
} as const;

export const QUEUE_CENTER_CONTRACT = contractDocument as unknown as ContractDocument;
export const QUEUE_CENTER_SCHEMA_VERSION = QUEUE_CENTER_CONTRACT.schema_version;
export const QUEUE_CENTER_HTTP_TRANSFER = QUEUE_CENTER_CONTRACT.http_transfer;
export const QUEUE_CENTER_ENDPOINTS = QUEUE_CENTER_CONTRACT.endpoints;
export const QUEUE_CENTER_RELAY = QUEUE_CENTER_CONTRACT.relay;
export type QueueCenterEndpointRole = keyof typeof contractDocument.endpoints;

/** Render one relay topic (machines verbatim, pair takes machine_id). */
export function relayTopic(
  kind: 'machines' | 'pair',
  tokens: Record<string, string | number> = {},
): string {
  let topic = QUEUE_CENTER_RELAY.topics[kind];
  for (const [key, value] of Object.entries(tokens)) {
    topic = topic.replace(`{${key}}`, String(value));
  }
  return topic;
}

/** Render one relay update type (request / response / roster). */
export function relayEvent(name: 'request' | 'response' | 'roster'): string {
  return QUEUE_CENTER_RELAY.events[name];
}

/**
 * Render one contract-owned Laravel endpoint path. Templates live in
 * config/queue_center_contract.json `endpoints`; token values are URL path
 * segments and are percent-encoded here.
 */
export function queueCenterEndpoint(
  role: QueueCenterEndpointRole,
  tokens: Record<string, string | number> = {},
): string {
  let path = QUEUE_CENTER_ENDPOINTS[role];
  for (const [key, value] of Object.entries(tokens)) {
    path = path.replace(`{${key}}`, encodeURIComponent(String(value)));
  }
  return path;
}
export const QUEUE_CENTER_REALTIME_EVENTS = QUEUE_CENTER_CONTRACT.realtime.events;
export const QUEUE_CENTER_DIFF_DELIVERY = QUEUE_CENTER_CONTRACT.diff_delivery;
export const QUEUE_CENTER_DELIVERY_RECEIPT = QUEUE_CENTER_CONTRACT.delivery_receipt;
export const QUEUE_CENTER_TASK_PROGRESS = QUEUE_CENTER_CONTRACT.task_contract.progress_stages;
export const QUEUE_CENTER_CONTROL_NAMES = QUEUE_CENTER_CONTRACT.control_names;
export const QUEUE_CENTER_CALLBACK_ROLES = QUEUE_CENTER_CONTRACT.callback_queue_roles;
export const QUEUE_CENTER_SCOPES = Object.keys(QUEUE_CENTER_CONTRACT.section_scopes) as QueueCenterScope[];
const GLOBAL_TASK_TYPE_DEFINITIONS = QUEUE_CENTER_CONTRACT.task_contract.task_types as GlobalTaskTypeDefinition[];
export const QUEUE_CENTER_CATEGORY_KEYS = QUEUE_CENTER_CONTRACT.categories.map((category) => category.key);
export const QUEUE_CENTER_CATEGORY_CATALOG = QUEUE_CENTER_CONTRACT.categories.map((category) => {
  const taskType = GLOBAL_TASK_TYPE_DEFINITIONS.find(
    (definition) => definition.key === category.laravel_task_type,
  );
  return {
    ...category,
    claimants: taskType?.claimants ?? (category.capability
      ? QUEUE_CENTER_CONTRACT.capability_claimants[category.capability] ?? []
      : [category.primary_handler]),
  };
});
export const GLOBAL_TASK_STATUSES_BY_ROLE = QUEUE_CENTER_CONTRACT.task_contract.statuses.values;
const taskStatusesForRoles = (roles: string[]): GlobalTaskStatus[] => (
  roles.map((role) => GLOBAL_TASK_STATUSES_BY_ROLE[role] ?? role)
);
export const GLOBAL_TASK_STATUSES = taskStatusesForRoles(QUEUE_CENTER_CONTRACT.task_contract.statuses.all);
export const GLOBAL_TASK_LIVE_STATUSES = taskStatusesForRoles(QUEUE_CENTER_CONTRACT.task_contract.statuses.live);
export const GLOBAL_TASK_TERMINAL_STATUSES = taskStatusesForRoles(QUEUE_CENTER_CONTRACT.task_contract.statuses.terminal);
export const GLOBAL_TASK_WORKER_RESULT_STATUSES = taskStatusesForRoles(
  QUEUE_CENTER_CONTRACT.task_contract.statuses.worker_reportable,
);
export const GLOBAL_TASK_EVENTS_BY_ROLE = QUEUE_CENTER_CONTRACT.task_contract.events.values;
export const GLOBAL_TASK_TERMINAL_EVENTS = QUEUE_CENTER_CONTRACT.task_contract.events.terminal.map(
  (role) => GLOBAL_TASK_EVENTS_BY_ROLE[role] ?? role,
);
export const GLOBAL_TASK_STREAM_EVENTS_BY_ROLE = QUEUE_CENTER_CONTRACT.task_contract.stream_events;
export const GLOBAL_TASK_EXECUTION_TYPES_BY_ROLE = QUEUE_CENTER_CONTRACT.task_contract.execution_types;
export const GLOBAL_TASK_EXECUTION_TYPES = Object.values(GLOBAL_TASK_EXECUTION_TYPES_BY_ROLE);
export const GLOBAL_TASK_CAPABILITIES = Object.keys(
  QUEUE_CENTER_CONTRACT.capability_claimants,
) as GlobalTaskCapability[];
export const GLOBAL_TASK_CAPABILITIES_BY_ROLE = Object.fromEntries(
  GLOBAL_TASK_CAPABILITIES.map((capability) => [capability, capability]),
) as Record<string, GlobalTaskCapability>;
export const GLOBAL_TASK_CAPABILITY_LABELS = QUEUE_CENTER_CONTRACT.task_contract.capability_labels;
export const GLOBAL_TASK_PRIORITIES = QUEUE_CENTER_CONTRACT.task_contract.priorities;
export const GLOBAL_TASK_LIMITS = QUEUE_CENTER_CONTRACT.task_contract.limits;
export const GLOBAL_TASK_CAPABILITY_SINGLE_LANES = QUEUE_CENTER_CONTRACT.task_contract.capability_single_lanes;
export const GLOBAL_TASK_FAST_LANE_CAPABILITIES = QUEUE_CENTER_CONTRACT.task_contract.fast_lane_capabilities;
export const GLOBAL_TASK_WIRE_SHAPES = QUEUE_CENTER_CONTRACT.task_contract.wire_shapes;
export const GLOBAL_TASK_TYPE_CATALOG = GLOBAL_TASK_TYPE_DEFINITIONS;
function buildGlobalTaskTypeIndex(
  definitions: GlobalTaskTypeDefinition[],
): Record<string, GlobalTaskTypeDefinition> {
  const index: Record<string, GlobalTaskTypeDefinition> = {};
  for (const definition of definitions) {
    if (definition.ordering !== 'queue_position' && definition.ordering !== 'priority') {
      throw new Error(`Queue Center task type has invalid ordering: ${definition.key}`);
    }
    for (const name of [definition.key, ...(definition.aliases ?? [])]) {
      const normalized = name.trim().toLowerCase();
      if (!normalized || index[normalized]) {
        throw new Error(`Queue Center task type name is duplicated: ${normalized}`);
      }
      index[normalized] = definition;
    }
  }
  return index;
}

export const GLOBAL_TASK_TYPE_BY_KEY = buildGlobalTaskTypeIndex(GLOBAL_TASK_TYPE_CATALOG);
export const QUEUE_CENTER_QUEUE_POSITION_CONTROLS = QUEUE_CENTER_CONTROL_NAMES.filter(
  (taskType) => GLOBAL_TASK_TYPE_BY_KEY[taskType]?.ordering === 'queue_position',
);
export const GLOBAL_QUEUE_POSITION_TASK_ALIASES = GLOBAL_TASK_TYPE_CATALOG
  .filter((definition) => definition.ordering === 'queue_position')
  .flatMap((definition) => definition.aliases ?? []);
export type GlobalQueuePositionTaskAlias = (typeof GLOBAL_QUEUE_POSITION_TASK_ALIASES)[number];
export const GLOBAL_TASK_FAST_PROMOTABLE_TYPES = GLOBAL_TASK_TYPE_CATALOG
  .filter((definition) => definition.fast_promotable === true)
  .map((definition) => definition.key);
export function getGlobalTaskTypeClaimants(
  taskType: string,
): Array<'pycore' | 'chrome' | 'laravel'> {
  const definition = getGlobalTaskTypeDefinition(taskType);
  if (!definition) return [];
  if (definition.claimants) return [...definition.claimants];
  return definition.capability
    ? [...(QUEUE_CENTER_CONTRACT.capability_claimants[definition.capability] ?? [])]
    : [];
}

export function getGlobalTaskTypesForClaimant(
  claimant: 'pycore' | 'chrome' | 'laravel',
  capability?: GlobalTaskCapability,
): string[] {
  return GLOBAL_TASK_TYPE_CATALOG
    .filter((definition) => !capability || definition.capability === capability)
    .filter((definition) => getGlobalTaskTypeClaimants(definition.key).includes(claimant))
    .map((definition) => definition.key);
}

export const GLOBAL_TASK_HISTORY_BUCKETS = QUEUE_CENTER_CONTRACT.task_contract.history_buckets.all;

function assertGlobalTaskWireDtoCoverage(): void {
  for (const [shape, dtoFields] of Object.entries(GLOBAL_TASK_WIRE_DTO_FIELDS)) {
    const contractFields = GLOBAL_TASK_WIRE_SHAPES[shape] ?? [];
    const matches = contractFields.length === dtoFields.length
      && contractFields.every((field, index) => field === dtoFields[index]);
    if (!matches) {
      throw new Error(`Queue Center DTO drift detected for wire shape: ${shape}`);
    }
  }
}

assertGlobalTaskWireDtoCoverage();

export function getGlobalTaskTypeDefinition(taskType: unknown): GlobalTaskTypeDefinition | null {
  const key = typeof taskType === 'string' ? taskType.trim().toLowerCase() : '';
  return GLOBAL_TASK_TYPE_BY_KEY[key] ?? null;
}

/**
 * Single ordering authority for every end: queue_position-ordered (Queue
 * Center audio) tasks sort and bump by Laravel head tickets only; all other
 * tasks keep the contract-defined numeric priority.
 */
export function getGlobalTaskOrdering(taskType: unknown): GlobalTaskOrdering {
  const definition = getGlobalTaskTypeDefinition(taskType);
  if (!definition) return 'priority';
  if (definition.ordering !== 'queue_position' && definition.ordering !== 'priority') {
    throw new Error(`Queue Center task type has invalid ordering: ${String(taskType ?? '')}`);
  }
  return definition.ordering;
}

export function isGlobalTaskQueuePositionOrdered(taskType: unknown): boolean {
  return getGlobalTaskOrdering(taskType) === 'queue_position';
}

export function getGlobalTaskOrderValue(
  task: GlobalTaskOrderingRecord,
): number {
  const field = getGlobalTaskOrdering(task.task_type);
  return Number(task[field] ?? 0);
}

export function compareGlobalTasksByContract(
  left: GlobalTaskOrderingRecord,
  right: GlobalTaskOrderingRecord,
): number {
  return getGlobalTaskOrderValue(right) - getGlobalTaskOrderValue(left);
}

export function getGlobalTaskPromptPayloadField(taskType: unknown): string {
  return getGlobalTaskTypeDefinition(taskType)?.prompt_payload_field ?? 'question';
}

/** Read prompt text by the central primary field with legacy payload fallbacks. */
export function getGlobalTaskPromptText(
  taskType: unknown,
  payload: GlobalTaskPayload,
): string {
  const fields = Array.from(new Set([
    getGlobalTaskPromptPayloadField(taskType),
    'text',
    'source_text',
    'question',
    'prompt',
  ]));
  for (const field of fields) {
    const value = payload[field];
    if (typeof value === 'string' && value.trim() !== '') return value;
  }
  return '';
}

export function normalizeGlobalTaskHistoryType(rawTaskType: unknown): string {
  const history = QUEUE_CENTER_CONTRACT.task_contract.history_buckets;
  const taskType = typeof rawTaskType === 'string' ? rawTaskType.trim().toLowerCase() : '';
  if (!taskType) return history.fallback;
  if (history.all.includes(taskType)) return taskType;
  if (history.exact_aliases[taskType]) return history.exact_aliases[taskType];

  for (const rule of history.token_rules) {
    const matchesAll = rule.all.every((token) => taskType.includes(token));
    const matchesAny = rule.any.length === 0 || rule.any.some((token) => taskType.includes(token));
    if (matchesAll && matchesAny) return rule.bucket;
  }
  return history.fallback;
}

export function isGlobalTaskStatus(value: unknown): value is GlobalTaskStatus {
  return typeof value === 'string' && GLOBAL_TASK_STATUSES.includes(value);
}

export function isGlobalTaskExecutionType(value: unknown): value is GlobalTaskExecutionType {
  return typeof value === 'string' && GLOBAL_TASK_EXECUTION_TYPES.includes(value);
}

export function isGlobalTaskCapability(value: unknown): value is GlobalTaskCapability {
  return typeof value === 'string' && GLOBAL_TASK_CAPABILITIES.includes(value);
}

const toNumber = (value: unknown): number => (
  typeof value === 'number' && Number.isFinite(value)
    ? value
    : typeof value === 'string' && value.trim() && Number.isFinite(Number(value))
      ? Number(value)
      : 0
);

const toBoolean = (value: unknown): boolean => {
  if (typeof value === 'boolean') return value;
  if (typeof value === 'number') return value !== 0;
  return typeof value === 'string' && ['1', 'true', 'yes', 'on'].includes(value.trim().toLowerCase());
};

const toNullableString = (value: unknown): string | null => (
  typeof value === 'string' && value.trim() ? value.trim() : null
);

export function isQueueCenterScope(value: unknown): value is QueueCenterScope {
  return typeof value === 'string' && QUEUE_CENTER_SCOPES.includes(value as QueueCenterScope);
}

export function buildEmptyQueueCenterSection(
  scope: QueueCenterScope,
  observedAt: string | null = null,
): QueueCenterSectionContract {
  const definition = QUEUE_CENTER_CONTRACT.section_scopes[scope];
  const defaults = QUEUE_CENTER_CONTRACT.section_contract_defaults;
  return {
    type: scope,
    category: definition.category,
    queue: { ...defaults.queue },
    worker: { ...defaults.worker },
    toggle: { ...defaults.toggle },
    lifecycle: defaults.lifecycle,
    error_code: defaults.error_code,
    last_error: defaults.last_error,
    observed_at: observedAt,
    age_s: defaults.age_s,
    stale: defaults.stale,
  };
}

export function buildDefaultQueueCenterSections(
  observedAt: string | null = null,
): Record<QueueCenterScope, QueueCenterSectionContract> {
  return Object.fromEntries(
    QUEUE_CENTER_SCOPES.map((scope) => [scope, buildEmptyQueueCenterSection(scope, observedAt)]),
  ) as Record<QueueCenterScope, QueueCenterSectionContract>;
}

/** Parse display-safe values for the browser exchange layer. */
export function normalizeQueueCenterSections(
  rawContracts: unknown,
  envelopeGeneratedAt: string | null,
): Record<QueueCenterScope, QueueCenterSectionContract> {
  const result = buildDefaultQueueCenterSections(envelopeGeneratedAt);
  if (!rawContracts || typeof rawContracts !== 'object') return result;

  Object.entries(rawContracts).forEach(([scopeKey, rawValue]) => {
    if (!isQueueCenterScope(scopeKey) || !rawValue || typeof rawValue !== 'object') return;
    const raw = rawValue as Record<string, unknown>;
    const queue = raw.queue && typeof raw.queue === 'object' ? raw.queue as Record<string, unknown> : {};
    const worker = raw.worker && typeof raw.worker === 'object' ? raw.worker as Record<string, unknown> : {};
    const toggle = raw.toggle && typeof raw.toggle === 'object' ? raw.toggle as Record<string, unknown> : {};
    const lifecycle = raw.lifecycle;
    result[scopeKey] = {
      type: scopeKey,
      category: toNullableString(raw.category) ?? result[scopeKey].category,
      queue: {
        pending: toNumber(queue.pending),
        processing: toNumber(queue.processing),
        leased: toNumber(queue.leased),
        total: toNumber(queue.total),
      },
      worker: {
        online: toBoolean(worker.online),
        claimed: toNumber(worker.claimed),
        ok: worker.ok == null ? null : toNumber(worker.ok),
        fail: worker.fail == null ? null : toNumber(worker.fail),
        last_heartbeat: toNullableString(worker.last_heartbeat),
      },
      toggle: {
        requested_by: toNullableString(toggle.requested_by),
        enabled: toBoolean(toggle.enabled),
        reason: toNullableString(toggle.reason),
        graceful_stop: toBoolean(toggle.graceful_stop),
        paused_by_user: toggle.paused_by_user == null ? null : toBoolean(toggle.paused_by_user),
      },
      lifecycle: lifecycle === 'off' || lifecycle === 'starting' || lifecycle === 'on' || lifecycle === 'error'
        ? lifecycle
        : 'off',
      error_code: toNullableString(raw.error_code),
      last_error: toNullableString(raw.last_error),
      observed_at: toNullableString(raw.observed_at),
      age_s: raw.age_s == null ? null : toNumber(raw.age_s),
      stale: toBoolean(raw.stale),
    };
  });

  return result;
}
